Content.Server/GameObjects/EntitySystems/GhostSystem.cs

```diff
@@ -1,27 +1,88 @@
+using System.Collections.Generic;
+using System.Linq;
+using Content.Server.GameObjects.Components.Markers;
 using Content.Server.GameObjects.Components.Mobs;
 using Content.Server.GameObjects.Components.Observer;
+using Content.Server.Players;
+using Content.Shared.GameObjects.EntitySystems;
 using JetBrains.Annotations;
+using Robust.Server.GameObjects;
+using Robust.Server.Player;
 using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Log;
+using Robust.Shared.Timing;
 
 namespace Content.Server.GameObjects.EntitySystems
 {
     [UsedImplicitly]
-    public class GhostSystem : EntitySystem
+    public class GhostSystem : SharedGhostSystem
     {
+        [Dependency] private readonly IGameTiming _gameTiming = default!;
+        [Dependency] private readonly IPlayerManager _playerManager = default!;
+
         public override void Initialize()
         {
             base.Initialize();
 
+            SubscribeLocalEvent<GhostComponent, ComponentStartup>(OnGhostStartup);
+            SubscribeLocalEvent<GhostComponent, ComponentShutdown>(OnGhostShutdown);
+
+            SubscribeLocalEvent<GhostComponent, ExaminedEvent>(OnGhostExamine);
+
             SubscribeLocalEvent<GhostComponent, MindRemovedMessage>(OnMindRemovedMessage);
             SubscribeLocalEvent<GhostComponent, MindUnvisitedMessage>(OnMindUnvisitedMessage);
+
+            SubscribeNetworkEvent<GhostWarpsRequestEvent>(OnGhostWarpsRequest);
+            SubscribeNetworkEvent<GhostReturnToBodyRequest>(OnGhostReturnToBodyRequest);
+            SubscribeNetworkEvent<GhostWarpToLocationRequestEvent>(OnGhostWarpToLocationRequest);
+            SubscribeNetworkEvent<GhostWarpToTargetRequestEvent>(OnGhostWarpToTargetRequest);
+        }
+
+        private void OnGhostStartup(EntityUid uid, GhostComponent component, ComponentStartup args)
+        {
+            // Allow this entity to be seen by other ghosts.
+            if (component.Owner.TryGetComponent(out VisibilityComponent? visibility))
+            {
+                visibility.Layer |= (int) VisibilityFlags.Ghost;
```

Comment:
This isn't your fault, but we need to unset `VisibilityFlags.Normal` (which is on by default) here too or ghost entity updates will be sent to alive players too, which is undesirable.

Reply:
Done in https://github.com/space-wizards/space-station-14/pull/4159/commits/d1d4589647b7c53883d0ac2fcf4b58d7bf5f7131